src/promise-pool-executor.ts

```diff
@@ -450,13 +461,24 @@ export class PromisePoolExecutor<T, R> implements UsesConcurrency, Stoppable, St
    * @returns {Promise}
    */
   async process (): Promise<ReturnValue<T, R>> {
-    for (const [index, item] of this.items().entries()) {
+    let index = 0
+    for await (const item of this.items()) {
       if (this.isStopped()) {
         break
       }
 
+      if (this.shouldUseCorrespondingResults()) {
+        this.results()[index] = PromisePool.notRun
+      }
+
       await this.waitForProcessingSlot()
```

Comment:
I guess we can remove this and only keep the `waitForProcessingSlot` at the end of this loop. This one doesn’t have any effect, does it?

Reply:
IIRC, this does nothing when `items` is an array, but if it's an iterable it's quite vital: imagine a scenario with iterable containing `1,2,3,4,5`, the pool starts processing items `1,2,3`, then `1` and `3` finish and then the user stops the pool. The result would be `[1, notRun, 3]`. If this code were to be removed, the second item would either be `undefined` or `<empty slot>` (not exactly sure how JS works here). I can add a comment explaining why the statement is needed, tho?

Reply:
This pull request changes the behavior of prefilling the results array to only prefill when items are an array. We’re not prefilling if the items come from an iterable. And we’re only prefilling for arrays if we want corresponding results.

I tested: tests are succeeding when removing this line which waits for a processing slot. Keeping the waiting line at the end though is crucial.